docs/msi_v2/msi_with_credential_design.md

```diff
@@ -0,0 +1,324 @@
+
+## Source Detection Logic
+
+MSAL follows a source detection process to determine how to interact with MSI endpoints and acquire tokens.
+
+### Environment Variable Check
+
+MSAL checks for Azure resource type based on specific environment variables to determine if the application is running on:
+
+- **Service Fabric**
+- **App Service**
+- **Machine Learning**
+- **Cloud Shell**
+- **Azure Arc**
+
+If identified, MSAL will use the appropriate legacy MSI endpoint for that resource.
+
+### Fallback to IMDS
+
+- If no specific Azure resource is identified from the environment variables, MSAL will fall back to IMDS (VMs and VMSS).
+- This fallback is the MSI v1 design or the legacy fallback mechanism.
+- In this new MSI v2 design, Before fully falling back to IMDS, MSAL will now **probe for the Credential Endpoint**.
+- MSAL probes to see if the `/credential` endpoint exists.
+- If the `/credential` endpoint is unavailable, it falls back to the legacy `/token` endpoint.
+- If probe is succesful then we can assume the current Azure Resource is a VM/VMSS 
+
+## MSI V2 /credential Endpoint Details
+
+### Short-Lived Credential Retrieval
+
+- The `/credential` endpoint provides a **temporary credential** instead of an access token.
+- This credential is only valid for a short duration (1 hour) and must be used **immediately** to acquire an access token from ESTS.
+- This mechanism improves security by reducing the lifetime of sensitive authentication materials.
+
+### Retry Logic
+
+MSAL uses the **default Managed Identity [retry](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/651b71c7d1dcaf3261e598e01e017dfd3672bb25/src/client/Microsoft.Identity.Client/Http/HttpManagerFactory.cs#L28) policy** for MSI V2 credential/token requests, whether calling the ESTS endpoint or the new `/credential` endpoint. i.e. MSAL performs 3 retries with a 1 second pause between each retry. Retries are performed on certain error [codes](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/651b71c7d1dcaf3261e598e01e017dfd3672bb25/src/client/Microsoft.Identity.Client/Http/HttpRetryCondition.cs#L12) only.
+
+## Steps for MSI V2 Authentication
+
+This section outlines the necessary steps to acquire an access token using the MSI V2 `/credential` endpoint. 
+
+### 1. Check for an Existing (Platform) Certificate (Windows only)
+- Search for a specific certificate (`devicecert.mtlsauth.local`) in `(Cert:\LocalMachine\My)`.
+- If the certificate is not found in Local Machine, check Current User's certificate store `(Cert:\CurrentUser\My)`.
+
+### 2. Generate a New Certificate (if platform certificate is not found)
+- If no valid platform certificate is found in Cert:\LocalMachine\My or Cert:\CurrentUser\My, create a new in-memory self-signed certificate.
+- This applies especially to Linux VMs, where platform certificates are not pre-configured, and MSAL must always generate an in-memory certificate for MTLS authentication.
```

Comment:
nit: "This will always apply to Linux VMs..."

Reply:
yes it has been mentioned